src/pubsub/node_pep.erl

```diff
@@ -165,17 +167,21 @@ get_affiliation(Nidx, Owner) ->
 set_affiliation(Nidx, Owner, Affiliation) ->
     node_flat:set_affiliation(Nidx, Owner, Affiliation).
 
-get_entity_subscriptions(Host, Owner) ->
-    {U, D, _} = SubKey = jid:to_lower(Owner),
-    GenKey = jid:to_bare(SubKey),
-    States = case SubKey of
-        GenKey ->
-            mnesia:match_object(#pubsub_state{stateid = {{U, D, '_'}, '_'}, _ = '_'});
-        _ ->
-            mnesia:match_object(#pubsub_state{stateid = {GenKey, '_'}, _ = '_'})
-            ++
-            mnesia:match_object(#pubsub_state{stateid = {SubKey, '_'}, _ = '_'})
-    end,
+get_entity_subscriptions(Host, #jid{ lserver = D, lresource = R } = Owner) ->
+    get_entity_subscriptions(Host, D, R, Owner);
+get_entity_subscriptions(Host, {_, D, R} = Owner) ->
+    get_entity_subscriptions(Host, D, R, Owner).
+
+get_entity_subscriptions(Host, D, R, Owner) ->
+    States = case R of
+                 <<>> ->
+                     {ok, States0} = mod_pubsub_db_backend:get_states_by_lus(Owner),
+                     States0;
+                 _ ->
+                     {ok, States0} = mod_pubsub_db_backend:get_states_by_bare(Owner),
+                     {ok, States1} = mod_pubsub_db_backend:get_states_by_full(Owner),
```

Comment:
As far as I can tell `get_states_by_full` are called only with `get_states_by_bare`. Maybe we could have `get_states_by_bare_and_full` instead of `get_states_by_full`. I'd leave `get_states_by_bare` as is because I saw it used alone in some places.
This may help optimize RDBMS backend, I'm not sure if Mnesia can be optimized to send only one query here.

Reply:
I was going to merge them if feasible during RDBMS backend implementation but you're right - there is a very good chance RDBMS may be optimised for such operation anyway.